Adding the MAV_GPS type

allanm84 opened the pull request:
This adds the MAV_GPS message which allows support for a GPS input from a companion computer.  Support in Ardupilot is being added with this PR: https://github.com/ArduPilot/ardupilot/pull/4178.  @tridge and @rmackay9 have been supporting this effort.

rmackay9 commented:
The advice from Tridge regarding a new "EXTERNAL_ABSOLUTE_POSITION" message was a bit different than what's in this PR.  So this was the advice:

EXTERNAL_ABSOLUTE_POSITION (as compared to HIL_GPS message):
- remove redundant cog field   <-- left in the MAV_GPS message
- add velocity accuracy estimate number  <-- ok!
- add position accuracy (ekf will be guessing)  <-- ok!
- make velocities floats in m/s  <-- left as int16t in MAV_GPS message

allanm84 commented:
@rmackay9 I thought about the "EXTERNAL_ABSOLUTE_POSITION" a bit, and this message really isn't that; it is intended to provide actual GPS data.  So, the fields that are in this message are meant to correspond to data that would be coming out of a GPS.  Thats why I left the COG/SOG values in there.  Also, RE the velocities: I was going to move them to float m/s values, but that would leave them inconsistent with the speed accuracy field (and therefore the other accuracy fields), which is in cm/s.

I am good with the GPS_MAV though, I pushed that change just now.

WickedShell commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,27 @@
+            <field type="uint8_t" name="fix_type">0-1: no fix, 2: 2D fix, 3: 3D fix. 4: 3D with DGPS. 5: 3D with RTK</field>
+            <field type="int32_t" name="lat">Latitude (WGS84), in degrees * 1E7</field>
+            <field type="int32_t" name="lon">Longitude (WGS84), in degrees * 1E7</field>
+            <field type="int32_t" name="alt">Altitude (AMSL, not WGS84), in cm (positive for up)</field>
```
The consensus in the past has been to move away from implementing these as int's and present it as a float in meters. (Int doesn't save any space, forces an arbitrary resolution on it, and we don't have round off problems with floats over the applicable range).

It's also worth noting that RTK GPS's are become available that do preform on the centimeter level already, so this is a problem for futureproofing reasons.

WickedShell commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,27 @@
+            <field type="uint32_t" name="time_week_ms">GPS time (milliseconds from start of GPS week)</field>
+            <field type="uint16_t" name="time_week">GPS week number</field>
+            <field type="uint8_t" name="fix_type">0-1: no fix, 2: 2D fix, 3: 3D fix. 4: 3D with DGPS. 5: 3D with RTK</field>
+            <field type="int32_t" name="lat">Latitude (WGS84), in degrees * 1E7</field>
```
Again from the perspective of future proof I think it's worth pointing out that again GPS unit's that are available today that according to the documentation are more accurate then this representation allows. However for lat/lon given how encoded this is elsewhere in MAVLink I don't think it makes sense to change it here. (IE I'd stick with the 1E7 representation you have).

rmackay9 commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,27 @@
+            <field type="uint16_t" name="vdop">GPS VDOP vertical dilution of position in cm</field>
+            <field type="int16_t" name="vn">GPS velocity in cm/s in NORTH direction in earth-fixed NED frame</field>
+            <field type="int16_t" name="ve">GPS velocity in cm/s in EAST direction in earth-fixed NED frame</field>
+            <field type="int16_t" name="vd">GPS velocity in cm/s in DOWN direction in earth-fixed NED frame</field>
```
I'm ok with sticking with int32 for the lat, lon but I really wonder about the velocity fields.  I think those should become floats.

WickedShell replied:
For what it's worth u-blox currently publishes this to us as mm/s (in a int32) so something better then cm/s would make sense to me. (I had the same thought about the heading, centidegrees seemed slight low res. (same with ground speed for that matter)

LorenzMeier replied:
Since this is used onboard there is indeed no win in using scaled integer here.

LorenzMeier commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,27 @@
+            <field type="int16_t" name="vd">GPS velocity in cm/s in DOWN direction in earth-fixed NED frame</field>
+            <field type="uint16_t" name="cog">Course over ground in centidegrees, 0..35999.</field>
+            <field type="int32_t" name="sog">Speed over ground in cm/s</field>
+            <field type="int32_t" name="speed_accuracy">GPS speed accuracy in cm/s</field>
```
These should be floats, as there is no win in using scaled integers here.

LorenzMeier commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,27 @@
             <field name="vert_accuracy" type="float">Vertical speed 1-STD accuracy</field>
         </message>
+        <message id="232" name="GPS_MAV">
+            <description>GPS sensor input message for the AP_GPS_MAV type</description>
```
The message is missing a GPS ID. This should be added so that data from multiple GPS units can be streamed.

allanm84 commented:
@LorenzMeier @rmackay9 @WickedShell I added an id field and changed all velocites, vdop/hdop, accuracies and cog/sog to floats in m and m/s.  I left the lat/lon/alt alone, as it wasn't clear to me from the comments if those should change to floats and/or lose their *1E7 scaling.  Please let me know if you think those should change.

WickedShell commented:
Alt should change, lat/lon is fine.
On May 22, 2016 6:14 PM, "Allan Matthew" <notifications@github.com> wrote:

> @LorenzMeier <https://github.com/LorenzMeier> @rmackay9
> <https://github.com/rmackay9> @WickedShell
> <https://github.com/WickedShell> I added an id field and changed all
> velocites, vdop/hdop, accuracies and cog/sog to floats in m and m/s. I left
> the lat/lon/alt alone, as it wasn't clear to me from the comments if those
> should change to floats and/or lose their *1E7 scaling. Please let me know
> if you think those should change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mavlink/mavlink/pull/555#issuecomment-220868894>
>

allanm84 commented:
@WickedShell thanks, changed.

rmackay9 commented:
Looks good to me.  i vote for it going in.  Perhaps as part of the merge @LorenzMeier or @tridge could swash the commits together.

allanm84 commented:
Squashed.

tridge commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,28 @@
+            <field type="float" name="vn">GPS velocity in m/s in NORTH direction in earth-fixed NED frame</field>
+            <field type="float" name="ve">GPS velocity in m/s in EAST direction in earth-fixed NED frame</field>
+            <field type="float" name="vd">GPS velocity in m/s in DOWN direction in earth-fixed NED frame</field>
+            <field type="float" name="cog">Course over ground in degrees, 0..359.99.</field>
```
I don't think we need cog and sog as we have vn, ve and vd. The velocity vector is a much better representation.

tridge commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,28 @@
+            <field type="float" name="speed_accuracy">GPS speed accuracy in m/s</field>
+            <field type="float" name="horiz_accuracy">GPS horizontal accuracy in m</field>
+            <field type="float" name="vert_accuracy">GPS vertical accuracy in m</field>
+            <field type="uint8_t" name="satellites_visible">Number of satellites visible.</field>
```
some GPS sources can't provide vertical velocity and the accuracy numbers. We've had trouble in the past with NMEA GPS modules and providing zero as vertical velocity.
I think we need a uint8_t flags field which says which of the optional fields are filled in.

tridge commented on message_definitions/v1.0/common.xml:
```diff
@@ -3329,6 +3329,28 @@
+        <message id="232" name="GPS_MAV">
+            <description>GPS sensor input message for the AP_GPS_MAV type</description>
+            <field type="uint8_t" name="gps_id">ID of the GPS for multiple GPS_MAV inputs</field>
+            <field type="uint16_t" name="ignore">Flags indicating which fields to ignore: bit0:alt, bit1:hdop, bit2:vdop, bit3:vel(vn+ve+vd), bit4:cog, bit5:sog, bit6:speed_accuracy, bit7:horiz_accuracy, bit8:vert_accuracy. All other fields must be provided.</field>
```
this has a single bit for all 3 velocity fields. Many GPS modules only provide horizontal velocity (vn and ve) not vd. I think we need separate bits for the horizontal and vertical velocity.
We should also remove the sog and cog, and the associated ignore bits.

tridge commented:
hi Allan, I've added a couple of inline comments above. Main thing is:

- remove cog and sog
- have separate ignore bit for horizontal and vertical velocity

apart from that this looks good!

magicrub commented:
lets not forget that right now the EKF2 takes into account a GPS latency.
This will complicate that

On Tue, May 24, 2016 at 6:06 PM, Andrew Tridgell <notifications@github.com>
wrote:

> hi Allan, I've added a couple of inline comments above. Main thing is:
>
>    - remove cog and sog
>    - have separate ignore bit for horizontal and vertical velocity
>
> apart from that this looks good!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mavlink/mavlink/pull/555#issuecomment-221446638>
>

allanm84 commented:
@tridge Thanks for the comments.  I gave in and got rid of COG/SOG and made the changes to the ignore flag you requested.

@magicrub Indeed.  Tridge and I chatted about that a bit, and it seems like the major issues to be worried about are:

1.) Proper timestamping of the packet
2.) Ensuring latency due to queueing in the serial buffers of the mavlink connection is minimized
3.) Good time synchronization between the companion computer and Pixhawk
4.) Measuring latency associated with this method of sending GPS information and adjusting the EK2_GPS_DELAY parameter (in APM)

What else needs to be considered?

rmackay9 commented:
You've run the gauntlet and come out alive.  Surely that's it.

LorenzMeier commented:
Looks pretty good now.

LorenzMeier commented:
Wait - I'm missing one bit: Why don't we just include a timestamp? We can sync the time across multiple systems. We could also allow 0 to indicate there is no valid timestamp.

allanm84 commented:
@LorenzMeier The motivation for using the GPS time format (week, week_ms) was that just a timestamp alone is ambiguous.  For instance, its not clear in the HIL_GPS message if the timestamp is intended to be GPS time or system time.

So, we could still add an additional timestamp, but it may be redundant to the GPS time variables.  Let me know what you think.

rmackay9 commented:
I've created a follow-up PR which I think addresses Lorenz's request that we add a time_usec field.  https://github.com/mavlink/mavlink/pull/589.

If that one goes in then this one can be closed.

LorenzMeier commented:
This is now on master

rmackay9 commented:
Great, thanks very much.
